plugins/generic/recommendBySimilarity/RecommendBySimilarityPlugin.inc.php

```diff
@@ -56,45 +62,87 @@ function getDescription() {
 		$articleId = $displayedArticle->getId();
 		import('classes.search.ArticleSearch');
+		import('classes.search.ArticleSearchIndex');
 		$articleSearch = new ArticleSearch();
-		$searchTerms = $articleSearch->getSimilarityTerms($articleId);
-		if (empty($searchTerms)) return false;
+		$keywords = (new ArticleSearchIndex())->filterKeywords($articleSearch->getSimilarityTerms($articleId));
+		$keywords = array_filter(array_unique($keywords), 'strlen');
+		if (!count($keywords)) {
+			return false;
+		}
 
-		// If we got similarity terms then execute a search with...
-		// ... request, journal and error messages, ...
 		$request = Application::get()->getRequest();
 		$router = $request->getRouter();
 		$journal = $router->getContext($request);
-		$error = null;
-		// ... search keywords ...
-		$query = implode(' ', $searchTerms);
-		$keywords = array(null => $query);
-		// ... and pagination.
 		$rangeInfo = Handler::getRangeInfo($request, 'articlesBySimilarity');
-		$rangeInfo->setCount(RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
-		$smarty->assign(array(
-			'articlesBySimilarity' => $articleSearch->retrieveResults(
-					$request,
-					$journal,
-					$keywords,
-					$error,
-					null, null,
-					$rangeInfo,
-					array($articleId)
-			),
-			'articlesBySimilarityQuery' => $query,
-		));
+		$rangeInfo->setCount(static::RECOMMEND_BY_SIMILARITY_PLUGIN_COUNT);
+
+		/** @var Builder */
+		$queryBuilder = (new SubmissionQueryBuilder())
+			->filterByContext($journal->getId())
+			->filterByStatus(STATUS_PUBLISHED)
+			->getQuery();
+
+		$queryBuilder->where('s.submission_id', '<>', $articleId);
+
+		$keywordIdFields = [];
+		$orderByMatches = $orderByMatchCount = [];
+		foreach ($keywords as $i => $term) {
+			$alias = "k{$i}";
+			$keywordIdFields[] = "{$alias}.keyword_id";
+			$queryBuilder->leftJoin("submission_search_keyword_list AS {$alias}", function (JoinClause $join) use ($term, $alias) {
+				$join->where("{$alias}.keyword_text", '=', $term);
+			});
+			$orderBy = '(' . Capsule::table('submission_search_objects', 'o')
+				->join("submission_search_object_keywords AS o{$i}", "o{$i}.object_id", '=', 'o.object_id')
+				->whereColumn("{$alias}.keyword_id", '=', "o{$i}.keyword_id")
+				->whereColumn('s.submission_id', '=', 'o.submission_id')
+				->selectRaw('COUNT(0)')
+				->toSql() . ')';
+			$orderByMatchCount[] = $orderBy;
+			$orderByMatches[] = "CASE WHEN {$orderBy} = 0 THEN 1 ELSE 0 END";
+		}
+		// Clear any previous ORDER BY
+		$queryBuilder->orders = [];
+		$queryBuilder->orderBy(Capsule::raw(implode(' + ', $orderByMatches)))
+			->orderByDesc(Capsule::raw(implode(' + ', $orderByMatchCount)))
+			->whereExists(function (Builder $query) use ($keywordIdFields) {
+				$query->from('submission_search_objects', 'o')
+					->join('submission_search_object_keywords AS ok', 'o.object_id', '=', 'ok.object_id')
+					->whereIn('ok.keyword_id', array_map('\\Illuminate\\Database\\Capsule\\Manager::raw', $keywordIdFields))
```

Comment:
It'll be nice to have access to [first-class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php) when we set our baseline at PHP8.1.